lib/internal/url.js

```diff
@@ -0,0 +1,328 @@
+        this[context].query = args[binding.ARG_QUERY];
+      // TODO: set the search params object also
+    });
+  }
+
+  get hash() {
+    return this[context].fragment === undefined ?
+      '' : `#${this[context].fragment}`;
+  }
+
+  set hash(hash) {
+    if (this.protocol === 'javascript:')
+      return;
+    if (!hash) {
+      delete this[context].fragment;
+      return;
+    }
+    if (typeof hash !== 'string')
+      throw new TypeError('hash must be a string');
+    if (hash[0] === '#') hash = hash.slice(1);
+    this[context].fragment = '';
+    binding.parse(hash, binding.kFragment, null, (...args) => {
+      // TODO: handle failure cases
+      if (args[binding.ARG_FRAGMENT])
+        this[context].fragment = args[binding.ARG_FRAGMENT];
+    });
+  }
+
+  get href() {
+    return this.toString();
+  }
+
+  toString(options) {
+    options = options || {};
+    const fragment =
+      options.fragment !== undefined ?
+        !!options.fragment : true;
+    const unicode = !!options.unicode;
+    var ret;
+    if (this.protocol)
+      ret = this.protocol;
+    if (this[context].host !== undefined) {
+      ret += '//';
+      if (this[context].username !== undefined ||
+          this[context].password !== undefined) {
+        if (this[context].username !== undefined)
+          ret += this[context].username;
+        if (this[context].password !== undefined)
+          ret += `:${this[context].password}`;
+        ret += '@';
+      }
+      if (unicode) {
+        ret += punycode.toUnicode(this.hostname);
+        if (this.port !== undefined)
+          ret += `:${this.port}`;
+      } else {
+        ret += this.host;
+      }
+    } else if (this[context].scheme === 'file:') {
+      ret += '//';
+    }
+    if (this.pathname)
+      ret += this.pathname;
+    if (this.search)
+      ret += this.search;
+    if (fragment)
+      ret += this.hash;
+    return ret;
+  }
+
+  inspect(depth, opts) {
+    var ret = 'URL {\n';
+    ret += `  href: ${this.href}\n`;
+    if (this.protocol !== undefined)
+      ret += `  protocol: ${this.protocol}\n`;
+    if (this.username !== undefined)
+      ret += `  username: ${this.username}\n`;
+    if (this.password !== undefined)
+      ret += `  password: ---------\n`;
+    if (this.hostname !== undefined)
+      ret += `  hostname: ${this.hostname}\n`;
+    if (this.port !== undefined)
+      ret += `  port: ${this.port}\n`;
+    if (this[context].path !== undefined)
+      ret += `  pathname: ${this.pathname}\n`;
+    if (this[context].query !== undefined)
+      ret += `  search: ${this.search}\n`;
+    if (this[context].fragment !== undefined)
+      ret += `  hash: ${this.hash}\n`;
+    if (opts.showHidden) {
+      ret += `  cannot-be-base: ${this[cannotBeBase]}\n`;
+    }
+    ret += '}';
+    return ret;
+  }
+}
+
+var hexTable = new Array(256);
+for (var i = 0; i < 256; ++i)
+  hexTable[i] = '%' + ((i < 16 ? '0' : '') + i.toString(16)).toUpperCase();
+function encodeAuth(str) {
```

Comment:
I don't believe this faithfully implements https://url.spec.whatwg.org/#userinfo-encode-set.